### ISSUE TITLE
Unexpectedly large datafile relative to amount of transfers and accounts

### ISSUE THREAD

**Issue by bakerg** [tigerbeetle/tigerbeetle#2039]
Hi there, we're currently assessing tigerbeetle and see that after creating around 250k accounts and 250k transfers the datafile is over 50GB in size. Based on the statement in https://docs.tigerbeetle.com/deploy/hardware that says
```
A 20 TiB disk containing a replica's data file is enough to address on the order of 50 billion accounts or transfers.
```
we were expecting each entity in tigerbeetle to require somewhere around 400bytes in the datafile, but in this instance it seems like approximately 100KB is used for each account or transfer. Is this the expected behaviour?

For the first ~100k accounts and transfers we were inserting one at a time to tigerbeetle and the rest after that were inserted in batches of 100 to increase the throughput. We have theorised that perhaps those inital inserts of single elements could've caused more space to be consumed in the datafile but it would be good to know for sure whether the input batch sizes are expected to affect the datafile size.

**Comment by cb22** [comment]
Hey @bakerg!

> For the first ~100k accounts and transfers we were inserting one at a time to tigerbeetle and the rest after that were inserted in batches of 100 to increase the throughput. We have theorised that perhaps those inital inserts of single elements could've caused more space to be consumed in the datafile but it would be good to know for sure whether the input batch sizes are expected to affect the datafile size.

Yes - that's the case currently. At the moment, there's heavy space amplification on smaller batch sizes; with anything below ~400 accounts or transfers in a batch experiencing it worse. (our focus for our first production release was on full batches)

We're working on something to significantly reduce the impact for our next release, though, which should be shortly after https://github.com/tigerbeetle/tigerbeetle/pull/2013 lands!

**Comment by bakerg** [comment]
Thanks for the quick response and that's good news about working on improving it. I'll run our test again with full batches and see what happens.

It might be helpful to mention that storage space is affected by batch size in the docs (maybe here? https://docs.tigerbeetle.com/reference/requests/#batching-events) as I think it currently only talks about throughput considerations.

**Comment by sentientwaffle** [comment]
As of https://github.com/tigerbeetle/tigerbeetle/pull/2070 (included in version 0.15.4), batch size no longer has a major impact on storage space used. In particular, the pathological space amplification of the one-transfer-per-batch case is fixed.
